File: StoreRoom/Common/Common/YCTools/ViewAutolayout/ViewAutolayoutCenter.m
//
//  ViewAutolayoutCenter.m
//  Common
//
//  Created by wlpiaoyi on 15/1/5.
//  Copyright (c) 2015年 wlpiaoyi. All rights reserved.
//

#import "ViewAutolayoutCenter.h"
#import "UIView+Expand.h"

@implementation ViewAutolayoutCenter

+(void) removeConstraints:(UIView*) subView{
    subView.translatesAutoresizingMaskIntoConstraints = NO;
    NSArray *constraints = [subView constraints];
    [subView removeConstraints:constraints];
    UIView *superView = [subView superview];
    constraints = [superView constraints];
    if (constraints&&[constraints count]) {
        for (NSLayoutConstraint *constraint in constraints) {
            if (constraint.secondItem==subView||constraint.firstItem==superView) {
                [superView removeConstraint:constraint];
            }
        }
    }
}


+(void) persistConstraintRelation:(UIView*) subView margins:(UIEdgeInsets) margins toItems:(NSDictionary*) toItems{
    
    subView.translatesAutoresizingMaskIntoConstraints = NO;
    if ([self isValueEnable:margins.top]) {
        UIView *superView = [toItems objectForKey:@"top"];
        NSLayoutAttribute superAtt = NSLayoutAttributeBottom;
        if (!superView) {
            superView = [subView superview];
            superAtt = NSLayoutAttributeTop;
        }
        NSLayoutConstraint *marginsTop = [NSLayoutConstraint constraintWithItem:subView attribute:NSLayoutAttributeTop relatedBy:NSLayoutRelationEqual toItem:superView attribute:superAtt multiplier:1 constant:margins.top];
        [[subView superview] addConstraint:marginsTop];
    }
    if ([self isValueEnable:margins.bottom]) {
        UIView *superView = [toItems objectForKey:@"bottom"];
        
        NSLayoutAttribute superAtt = NSLayoutAttributeTop;
        if (!superView) {
            superView = [subView superview];
            superAtt = NSLayoutAttributeBottom;
        }
        NSLayoutConstraint *marginsBottom = [NSLayoutConstraint constraintWithItem:subView attribute:NSLayoutAttributeBottom relatedBy:NSLayoutRelationEqual toItem:superView attribute:superAtt multiplier:1 constant:-margins.bottom];
        [[subView superview] addConstraint:marginsBottom];
    }
    if ([self isValueEnable:margins.left]) {
        UIView *superView = [toItems objectForKey:@"left"];
        NSLayoutAttribute superAtt = NSLayoutAttributeRight;
        if (!superView) {
            superView = [subView superview];
            superAtt = NSLayoutAttributeLeft;
        }
        NSLayoutConstraint *marginsLeft = [NSLayoutConstraint constraintWithItem:subView attribute:NSLayoutAttributeLeft relatedBy:NSLayoutRelationEqual toItem:superView attribute:superAtt multiplier:1 constant:margins.left];
        [[subView superview] addConstraint:marginsLeft];
    }
    if ([self isValueEnable:margins.right]) {
        UIView *superView = [toItems objectForKey:@"right"];
        NSLayoutAttribute superAtt = NSLayoutAttributeLeft;
        if (!superView) {
            superView = [subView superview];
            superAtt = NSLayoutAttributeRight;
        }
        NSLayoutConstraint *marginsRight = [NSLayoutConstraint constraintWithItem:subView attribute:NSLayoutAttributeRight relatedBy:NSLayoutRelationEqual toItem:superView attribute:superAtt multiplier:1 constant:-margins.right];
        [[subView superview] addConstraint:marginsRight];
    }
}

+(void) persistConstraintSize:(UIView*) subView{
    subView.translatesAutoresizingMaskIntoConstraints = NO;
    CGSize size = subView.frameSize;
    NSMutableArray *constraints = [NSMutableArray new];
    if ([self isValueEnable:size.width]) {
        NSLayoutConstraint *sizeWith = [NSLayoutConstraint constraintWithItem:subView attribute:NSLayoutAttributeWidth relatedBy:NSLayoutRelationEqual toItem:nil attribute:NSLayoutAttributeNotAnAttribute multiplier:1 constant:size.width];
        [constraints addObject:sizeWith];
    }
    if ([self isValueEnable:size.height]) {
        NSLayoutConstraint *sizeHeight= [NSLayoutConstraint constraintWithItem:subView attribute:NSLayoutAttributeHeight relatedBy:NSLayoutRelationEqual toItem:nil attribute:NSLayoutAttributeNotAnAttribute multiplier:1 constant:size.height];
        [constraints addObject:sizeHeight];
    }
    [subView addConstraints:constraints];
    
}
+(void) persistConstraintCenter:(UIView*) subView{
    subView.translatesAutoresizingMaskIntoConstraints = NO;
    UIView *superView = [subView superview];
    CGPoint offPoint = subView.frameOrigin;
    NSMutableArray *constraints = [NSMutableArray new];
    if ([self isValueEnable:offPoint.x]) {
        NSLayoutConstraint *centerX = [NSLayoutConstraint constraintWithItem:subView attribute:NSLayoutAttributeCenterX relatedBy:NSLayoutRelationEqual toItem:superView attribute:NSLayoutAttributeCenterX multiplier:1 constant:offPoint.x];
        [constraints addObject:centerX];
    }
    if ([self isValueEnable:offPoint.y]) {
        NSLayoutConstraint *centerY = [NSLayoutConstraint constraintWithItem:subView attribute:NSLayoutAttributeCenterY relatedBy:NSLayoutRelationEqual toItem:superView attribute:NSLayoutAttributeCenterY multiplier:1 constant:offPoint.y];
        [constraints addObject:centerY];
    }
    [superView addConstraints:constraints];
}

+(BOOL) isValueEnable:(float) value{
    if (value<DisableConstrainsValueMAX-1&&value>=DisableConstrainsValueMIN+1) {
        return YES;
    }
    return NO;
}

@end

File: StoreRoom/zhengweiurl1/zhengweiurl/ViewAutolayoutCenter.swift
//
//  ViewAutolayoutCenter.swift
//  zhengweiurl
//
//  Created by wlpiaoyi on 15/2/11.
//  Copyright (c) 2015年 wlpiaoyi. All rights reserved.
//

import UIKit
/**
 最大失效值
 */
let DisableConstrainsValueMAX:CGFloat = CGFloat.max/2;
/**
 最小失效值
 */
let DisableConstrainsValueMIN:CGFloat = -DisableConstrainsValueMAX;
/**
 自动布局设置
 */
class ViewAutolayoutCenter: NSObject {
    
//    class func aatest(function:functionTypetest) {
//        function(["a":"v","b":"v"]);
//    }
    
    /**
    删除相关的约束
    */
    class func removeConstraints(targetView:UIView){
//        self.aatest { (b) -> Void in
//            var v1 = b?.objectForKey("a");
//            var v2 = b?.objectForKey("b");
//        }
        targetView.setTranslatesAutoresizingMaskIntoConstraints(false);
        var constraints:NSArray? = targetView.constraints();
        if(constraints?.count > 0){
            for constraint in (constraints)! {
                if(constraint.isKindOfClass(NSLayoutConstraint.classForCoder())){
                    targetView.removeConstraint((constraint) as NSLayoutConstraint);
                    continue;
                }
                if(constraint.isKindOfClass(NSArray.classForCoder())){
                    targetView.removeConstraints((constraint) as NSArray);
                    continue;
                }
                
            }
        }
    }
    /**
    新增关系约束
    */
    class func persistConstraintRelation(targetView:UIView,margins:UIEdgeInsets,toItems:NSDictionary?){
        targetView.setTranslatesAutoresizingMaskIntoConstraints(false);
        if(self.isValueEnable(margins.top)){
            var itemView:UIView? = toItems?.objectForKey("top") as? UIView;
            var itemAtt:NSLayoutAttribute = NSLayoutAttribute.Bottom;
            var targetAtt:NSLayoutAttribute = NSLayoutAttribute.Top;
            var relation:NSLayoutRelation = NSLayoutRelation.Equal;
            var constant:CGFloat = margins.top;
            if(itemView == nil){
                itemView = targetView.superview;
                itemAtt = NSLayoutAttribute.Top;
            }
            var layoutConstraint:NSLayoutConstraint = NSLayoutConstraint(item: targetView, attribute: targetAtt, relatedBy: relation, toItem: itemView, attribute: itemAtt, multiplier: 1, constant: constant);
            targetView.superview?.addConstraint(layoutConstraint);
        }
        if(self.isValueEnable(margins.bottom)){
            var itemView:UIView? = toItems?.objectForKey("bottom") as? UIView;
            var itemAtt:NSLayoutAttribute = NSLayoutAttribute.Top;
            var targetAtt:NSLayoutAttribute = NSLayoutAttribute.Bottom;
            var relation:NSLayoutRelation = NSLayoutRelation.Equal;
            var constant:CGFloat = margins.bottom;
            if(itemView == nil){
                itemView = targetView.superview;
                itemAtt = NSLayoutAttribute.Bottom;
            }
            var layoutConstraint:NSLayoutConstraint = NSLayoutConstraint(item: targetView, attribute: targetAtt, relatedBy: relation, toItem: itemView, attribute: itemAtt, multiplier: 1, constant: -constant);
            targetView.superview?.addConstraint(layoutConstraint);
        }
        if(self.isValueEnable(margins.left)){
            var itemView:UIView? = toItems?.objectForKey("left") as? UIView;
            var itemAtt:NSLayoutAttribute = NSLayoutAttribute.Right;
            var targetAtt:NSLayoutAttribute = NSLayoutAttribute.Left;
            var relation:NSLayoutRelation = NSLayoutRelation.Equal;
            var constant:CGFloat = margins.left;
            if(itemView == nil){
                itemView = targetView.superview;
                itemAtt = NSLayoutAttribute.Left;
            }
            var layoutConstraint:NSLayoutConstraint = NSLayoutConstraint(item: targetView, attribute: targetAtt, relatedBy: relation, toItem: itemView, attribute: itemAtt, multiplier: 1, constant: constant);
            targetView.superview?.addConstraint(layoutConstraint);
        }
        if(self.isValueEnable(margins.right)){
            var itemView:UIView? = toItems?.objectForKey("right") as? UIView;
            var itemAtt:NSLayoutAttribute = NSLayoutAttribute.Left;
            var targetAtt:NSLayoutAttribute = NSLayoutAttribute.Right;
            var relation:NSLayoutRelation = NSLayoutRelation.Equal;
            var constant:CGFloat = margins.right;
            if(itemView == nil){
                itemView = targetView.superview;
                itemAtt = NSLayoutAttribute.Right;
            }
            var layoutConstraint:NSLayoutConstraint = NSLayoutConstraint(item: targetView, attribute: targetAtt, relatedBy: relation, toItem: itemView, attribute: itemAtt, multiplier: 1, constant: -constant);
            targetView.superview?.addConstraint(layoutConstraint);
        }
    }
    /**
    新增大小约束
    */
    class func persistConstraintSize(targetView:UIView,width:CGFloat,height:CGFloat){
        var layoutConstraints:NSMutableArray! = NSMutableArray();
        targetView.setTranslatesAutoresizingMaskIntoConstraints(false);
        if(self.isValueEnable(width)){
            var layoutConstraint:NSLayoutConstraint = NSLayoutConstraint(item: targetView, attribute: NSLayoutAttribute.Width, relatedBy: NSLayoutRelation.Equal, toItem: nil, attribute: NSLayoutAttribute.NotAnAttribute, multiplier:1, constant: width);
            layoutConstraints.addObject(layoutConstraint);
        }
        
        if(self.isValueEnable(height)){
            var layoutConstraint:NSLayoutConstraint = NSLayoutConstraint(item: targetView, attribute: NSLayoutAttribute.Height, relatedBy: NSLayoutRelation.Equal, toItem: nil, attribute: NSLayoutAttribute.NotAnAttribute, multiplier:1, constant: height);
            layoutConstraints.addObject(layoutConstraint);
            
        }
        targetView.superview?.addConstraints(layoutConstraints);
    }
    /*
    新增布局约束
    */
    class func persistConstraintCenter(targetView:UIView!, offsetx:CGFloat, offsety:CGFloat){
        var layoutConstraints:NSMutableArray! = NSMutableArray();
        targetView.setTranslatesAutoresizingMaskIntoConstraints(false);
        if(self.isValueEnable(offsetx)){
            var layoutConstraint:NSLayoutConstraint = NSLayoutConstraint(item: targetView, attribute: NSLayoutAttribute.CenterX, relatedBy: NSLayoutRelation.Equal, toItem: targetView.superview, attribute: NSLayoutAttribute.CenterX, multiplier:1, constant: offsetx);
            layoutConstraints.addObject(layoutConstraint);
        }
        if(self.isValueEnable(offsety)){
            var layoutConstraint:NSLayoutConstraint = NSLayoutConstraint(item: targetView, attribute: NSLayoutAttribute.CenterY, relatedBy: NSLayoutRelation.Equal, toItem: targetView.superview, attribute: NSLayoutAttribute.CenterY, multiplier:1, constant: offsety);
            layoutConstraints.addObject(layoutConstraint);
        }
        targetView.superview?.addConstraints(layoutConstraints);
    }
    /**
    检测值是否有效
    */
    private class func isValueEnable(value:CGFloat)->Bool{
        if(value<DisableConstrainsValueMAX-1&&value>=DisableConstrainsValueMIN+1){
            return true;
        }
        return false;
    }
   
}
